File: Algorithms/Searching/Java/LinearSearch.java
package Algorithms.Searching.Java;

import java.util.Arrays;
import java.util.Scanner;

// Time complexity of LinearSearch is O(n)

public class LinearSearch {

    static int[] arr = {2, 3, 4, 10, 40};
    static int x;
    static int position = -1;

    public static void main(String[] args) {

        System.out.println(Arrays.toString(arr));

        Scanner sc = new Scanner(System.in);
        System.out.println("Enter num to search in array");
        x = sc.nextInt();

        // int pos = findElement(x);

        int pos = improvedSearch(x, 0, (arr.length - 1));

        if (pos == -1) {
            System.out.println("Element not found");
        } else {
            System.out.println("Element found at position " + (pos + 1));
        }
    }

    /**
     * @param num
     * @return
     */
    public static int search(int num) {

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == num) {
                return i;
            }
        }

        return -1;
    }

    // Improve Linear Search Worst-Case Complexity

    public static int improvedSearch(int num, int left, int right) {

        for (left = 0; left <= right; ) {

            // if search_element is found with left variable
            if (arr[left] == num) {
                position = left;
                return position;
            }

            // if search_element is found with right variable

            if (arr[right] == num) {
                position = right;
                return position;
            }

            left++;
            right--;

        }

        // if element not found
        if (position == -1) {
            System.out.println("Not found in Array with "
                    + left + " Attempt");
        }

        return position;
    }

}

File: Algorithms/Searching/Python/LinearSearch.py

def search(arr, search_Element):
    for i in range(len(arr)):
        if arr[i] == search_Element:
            print("Element found at position " + str(i + 1))
            return

    print("Element not found")


def improvedSearch(arr, search_Element):
    left = 0
    length = len(arr)
    position = -1
    right = length - 1

    # Run loop from 0 to right
    for left in range(0, right, 1):

        # If search_element is found with
        # left variable
        if arr[left] == search_Element:
            position = left
            print("Element found in Array at ", position +
                  1, " Position with ", left + 1, " Attempt")
            break

        # If search_element is found with
        # right variable
        if arr[right] == search_Element:
            position = right
            print("Element found in Array at ", position + 1,
                  " Position with ", length - right, " Attempt")
            break
        left += 1
        right -= 1

    # If element not found
    if position == -1:
        print("Not found in Array with ", left, " Attempt")


# Driver code
arr = [1, 2, 3, 4, 5]
search_element = input("Enter an num to search in array")

# Function call
search(arr, int(search_element))
